### ISSUE TITLE
update_electricity() incorrect in SSP2

### ISSUE THREAD

**Issue by igt-misc** [polca/premise#145]
The update_electricity() function in 1.8.2.dev7 is producing unusual electricity markets for all years in all SSPs, which can be most easily seen when comparing all SSPs to one another and baseline ecoinvent 3.9.1 consequential.
![glo_LCA-scores_ReCiPe-2016-v103-midpoint-H-climate-change-global-warming-potential-GWP1000](https://github.com/polca/premise/assets/143397697/527d02a4-851d-4673-9e41-857ce151002d)

If we go into each SSP's database, some weird processes pop up:

### SSP1-Base 2030 Market group for high voltage electricity
![ssp1_30_mgehv](https://github.com/polca/premise/assets/143397697/d76dda9f-cc45-4c69-ae33-f30bb427d4a2)

### SSP2-Base 2030 Market group for high voltage electricity
![ssp2_30_mgehv](https://github.com/polca/premise/assets/143397697/8aad71c8-644f-4342-b11f-3a8529f481e1)

### SSP5-Base 2030 Market group for high voltage electricity
![ssp5_30_mgehv](https://github.com/polca/premise/assets/143397697/732e23d4-18b7-4d01-8357-f6f4f1a09647)

**Comment by igt-misc** [comment]
Addendum on the title of this issue, when I started writing this I thought it was just SSP2 but in hindsight it does in fact appear to apply across the board.

**Comment by romainsacchi** [comment]
Can you paste the script you ran?

**Comment by igt-misc** [comment]
scenaria = [
		{"model":"remind","pathway":"SSP1-Base","year":2030},
		{"model":"remind","pathway":"SSP2-Base","year":2030},
		{"model":"remind","pathway":"SSP5-Base","year":2030},
		{"model":"remind","pathway":"SSP1-Base","year":2040},
		{"model":"remind","pathway":"SSP2-Base","year":2040},
		{"model":"remind","pathway":"SSP5-Base","year":2040},
		{"model":"remind","pathway":"SSP1-Base","year":2050},
		{"model":"remind","pathway":"SSP2-Base","year":2050},
		{"model":"remind","pathway":"SSP5-Base","year":2050}]

ndb = premise.NewDatabase(
				scenarios=scenaria[0],
				source_db="ei_391_c",  # <-- Must point to the consequential database.
				source_version="3.9.1",
				key="tUePmX_S5B8ieZkkM7WUU2CnO8SmShwmAeWK9x2rTFo=",
				system_model="consequential",  # <-- Must specify "consequential"
				quiet=False,
				use_multiprocessing=True,
			)
ndb.update_electricity()
ndb.write_db_to_brightway(names)
ndb.write_superstructure_db_to_brightway()


Incidentally, in the last issuse you helped me with you can see this exact weirdness in the charts you posted as well [here](https://github.com/polca/premise/issues/134#issuecomment-1908622518) and [here](https://github.com/polca/premise/issues/134#issuecomment-1908563775). Compared to themselves, the impacts go down over years which makes sense. However, you'll notice that the SSP1 2020 is up at .8 kg CO2e per 1 kWh medium voltage, while SSP2 is at .14 kgCO2e for the same time period.

**Comment by romainsacchi** [comment]
Some data before delving into the issue.

The electricity production mixes as indicated in the REMIND SSP1, SSP2 and SSP2 scenarios.
At face value, I'd say that the volume trend is decreasing in SSP1 while increasing in SSP2 and SSP5. The market volume trend will influence the choice of marginal suppliers. Hence, the logic used to pick the marginal suppliers in SSP1 will be different from that of SSP2 and 5.


<img width="1391" alt="Screenshot 2024-01-31 at 10 37 08" src="https://github.com/polca/premise/assets/22473067/b987c8e5-9daa-4eab-a154-1d5539be129c">

**Comment by romainsacchi** [comment]
Now, here are the marginal mixes calculated by premise, still for EUR.
There is no doubt that the GWP values you obtained correspond to the mixes we see down here. Hence, there is no error between this point and the LCIA step. If there is an error, it's in the definition of the marginal mix itself.

<img width="389" alt="Screenshot 2024-01-31 at 10 45 33" src="https://github.com/polca/premise/assets/22473067/7fee8551-7949-4d8e-bee1-dd1d2e041e37">

**Comment by romainsacchi** [comment]
We need to look into [the marginal mix algorithm](https://github.com/polca/premise/blob/master/premise/marginal_mixes.py) to see how premise comes up with these mixes.

**Comment by romainsacchi** [comment]
When I run the SSP1 scenario, for EUR, the following parameters are used to calculate the marginal mix:

````python
##################################################
range time False
duration False
foresight False
lead time False
capital_repl_rate False
measurement 0
weighted_slope_start 0.75
weighted_slope_end 1.0
avg_start 2030
avg_end 2074
avg_lifetime 44
volume change -0.05395457899753466
##################################################
`````

So, we indeed have a decreasing market trend, and the algorithm is going to look at the production volumes of each tech over the next 44 years (which is the average lifetime of the mix) using the method 0 (i.e., the slope of each technology's production volume). With the default parameters value, the renewal rate of capital replacement is not considered. See [docs](https://premise.readthedocs.io/en/latest/consequential.html).

**Comment by romainsacchi** [comment]
So it proceeds with calculating the slopes of each tech (method 0, without capital renewal rate), which is this:

````python
variables            year              
Storage, Hydrogen    2030    EUR  0.000
Gas CHP              2030    EUR  0.000
Coal CHP             2030    EUR  0.000
Hydro                2030    EUR -0.007
Nuclear              2030    EUR -0.113
Gas CC CCS           2030    EUR  0.000
Coal IGCC CCS        2030    EUR  0.000
Biomass IGCC         2030    EUR -0.005
Gas ST               2030    EUR  0.002
Coal IGCC            2030    EUR  0.000
Coal PC              2030    EUR -0.035
Geothermal           2030    EUR  0.000
Wind Offshore        2030    EUR  0.017
Solar PV Centralized 2030    EUR  0.191
Gas CC               2030    EUR -0.080
Wind Onshore         2030    EUR  0.090
Biomass CHP          2030    EUR  0.000
Biomass IGCC CCS     2030    EUR  0.000
Oil ST               2030    EUR -0.010
Solar CSP            2030    EUR -0.001
Coal PC CCS          2030    EUR  0.000
````

**Comment by romainsacchi** [comment]
Then, because `volume change -0.05395457899753466`, it filters out positively trending techs, to keep only the negative ones and flips their sign:

````python
variables            year                 
Coal IGCC CCS        2030    EUR -0.000000
Gas CHP              2030    EUR -0.000000
Coal IGCC            2030    EUR -0.000000
Gas ST               2030    EUR -0.000000
Wind Onshore         2030    EUR -0.000000
Nuclear              2030    EUR  0.450199
Gas CC CCS           2030    EUR -0.000000
Gas CC               2030    EUR  0.318725
Biomass IGCC CCS     2030    EUR -0.000000
Coal PC              2030    EUR  0.139442
Coal PC CCS          2030    EUR -0.000000
Biomass IGCC         2030    EUR  0.019920
Solar PV Centralized 2030    EUR -0.000000
Coal CHP             2030    EUR -0.000000
Storage, Hydrogen    2030    EUR -0.000000
Biomass CHP          2030    EUR -0.000000
Oil ST               2030    EUR  0.039841
Solar CSP            2030    EUR  0.003984
Hydro                2030    EUR  0.027888
Wind Offshore        2030    EUR -0.000000
Geothermal           2030    EUR -0.000000
````

**Comment by romainsacchi** [comment]
The logic for this is the following:

````python
# market decreasing faster than the average capital renewal rate. <- equals to zero in our case
# in this case, the idea is that oldest/non-competitive technologies
# are likely to supply by increasing their lifetime
# as the market does not justify additional capacity installation
````
this is also explained in the paper https://doi.org/10.1016/j.rser.2023.113830

**Comment by romainsacchi** [comment]
I just realized there was an issue with calculating the renewal rate of capital replacement (which does not affect your case, since your parameter for "capital replacement rate" is set to False, as is its default value). But after fixing this issue, setting "capital replacement rate" to True modifies *slightly* the mixes:

<img width="384" alt="Screenshot 2024-01-31 at 11 45 37" src="https://github.com/polca/premise/assets/22473067/b992cc0a-79a4-4401-b674-1fa861369611">

Setting "capital replacement rate" to True means that a capital replacement rate will be calculated for each technology and subtracted from their respective slope to isolate the (de)growth due to actual capacity installed/removed from the mere fact of renewing aging infrastructure.

I think this should default to True, not False, as is currently the case. In any case, it does not significantly change your results.

**Comment by romainsacchi** [comment]
So, as far as I can tell, besides the renewal capital rate, which was miscalculated (but which did not enter your calculations), things do not really look incorrect. I think it is mostly due to the fact that the market is decreasing, leading to picking the least-preferred tech instead (because we assume economics dictate that those will go through lifetime extension instead of investing in new capacity in renewables on a market that does not offer the prospect of growth).

I will push the fix for the rate of capital renewal in `1.8.2.dev8`.
Also, I suggest you play with other parameters and methods to see how that changes the mix.

**Comment by romainsacchi** [comment]
As for the SSP5 scenario, the market is indeed increasing unlike SSP1, but we can see from the graphs above that one of the fastest-growing techs is Gas combined cycle ("Gas CC"), which explains its share in the mix (and the GWP). So, it makes sense (SSP5 is fossil fuels-driven).

<img width="452" alt="Screenshot 2024-01-31 at 12 04 28" src="https://github.com/polca/premise/assets/22473067/3380f996-fa6d-4aaa-abf0-fecf2da33583">
